pinot-core/src/main/java/org/apache/pinot/core/util/ListenerConfigUtil.java

```diff
@@ -0,0 +1,206 @@
+    if (queryPortString != null) {
+      listeners.add(new ListenerConfig(CommonConstants.HTTP_PROTOCOL, DEFAULT_HOST, Integer.parseInt(queryPortString),
+          CommonConstants.HTTP_PROTOCOL, new TlsConfig()));
+    }
+
+    TlsConfig tlsDefaults = TlsUtils.extractTlsConfig(brokerConf, CommonConstants.Broker.BROKER_TLS_PREFIX);
+
+    listeners.addAll(buildListenerConfigs(brokerConf, "pinot.broker.client", tlsDefaults));
+
+    return listeners;
+  }
+
+  public static List<ListenerConfig> buildServerAdminConfigs(PinotConfiguration serverConf) {
+    List<ListenerConfig> listeners = new ArrayList<>();
+
+    String adminApiPortString = serverConf.getProperty(CommonConstants.Server.CONFIG_OF_ADMIN_API_PORT);
+    if (adminApiPortString != null) {
+      listeners.add(new ListenerConfig(CommonConstants.HTTP_PROTOCOL, DEFAULT_HOST, Integer.parseInt(adminApiPortString),
+          CommonConstants.HTTP_PROTOCOL, new TlsConfig()));
+    }
+
+    TlsConfig tlsDefaults = TlsUtils.extractTlsConfig(serverConf, CommonConstants.Server.SERVER_TLS_PREFIX);
+
+    listeners.addAll(buildListenerConfigs(serverConf, "pinot.server.adminapi", tlsDefaults));
+
+    return listeners;
+  }
+
+  private static ListenerConfig buildListenerConfig(PinotConfiguration config, String namespace, String protocol,
+      TlsConfig tlsDefaults) {
+    String protocolNamespace = namespace + DOT_ACCESS_PROTOCOLS + "." + protocol;
+
+    TlsConfig tlsConfig = TlsUtils.extractTlsConfig(tlsDefaults, config, protocolNamespace + ".tls");
+    tlsConfig.setEnabled(CommonConstants.HTTPS_PROTOCOL.equals(protocol));
+
+    return new ListenerConfig(protocol,
+        getHost(config.getProperty(protocolNamespace + ".host", DEFAULT_HOST)),
+        getPort(config.getProperty(protocolNamespace + ".port")), protocol, tlsConfig);
+  }
+
+  private static String getHost(String configuredHost) {
+    return Optional.ofNullable(configuredHost).filter(host -> !host.trim().isEmpty())
+        .orElseThrow(() -> new IllegalArgumentException(configuredHost + " is not a valid host"));
+  }
+
+  private static int getPort(String configuredPort) {
+    return Optional.ofNullable(configuredPort).filter(port -> !port.trim().isEmpty()).<Integer> map(Integer::valueOf)
+        .orElseThrow(() -> new IllegalArgumentException(configuredPort + " is not a valid port"));
+  }
+
+  public static HttpServer buildHttpServer(ResourceConfig resConfig, List<ListenerConfig> listenerConfigs) {
+    Preconditions.checkNotNull(listenerConfigs);
+
+    // The URI is irrelevant since the default listener will be manually rewritten.
+    HttpServer httpServer = GrizzlyHttpServerFactory.createHttpServer(URI.create("http://0.0.0.0/"), resConfig, false);
+
+    // Listeners cannot be configured with the factory. Manual overrides are required as instructed by Javadoc.
+    httpServer.removeListener("grizzly");
+
+    listenerConfigs.forEach(listenerConfig -> configureListener(httpServer, listenerConfig));
+
+    try {
```

Comment:
Do we really need to start the http server here? The caller may choose to build the server ahead of time, and then start it in its own start process. I suggest we let the caller start the server.

Reply:
changed